internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,84 @@
+	Reincludes []string // Relative path under source to reinclude files that are excluded in the upload.
+	Excludes   []string // Relative path under source to exclude in the upload.
+	Recursive  bool     // Whether to walk recursively.
+}
+
+// Cached represents an S3 object uploaded to a cache bucket that needs
+// to be moved from a cached location to the destination bucket/key.
+type Cached struct {
+	// LocalPath is the local path to the asset.
+	LocalPath string
+
+	// Content is the content of the file at LocalPath.
+	Content io.Reader
+
+	// CachePath is the uploaded location of the asset in the CacheBucket.
+	CachePath string
+
+	// CacheBucket is the bucket the file was uploaded to.
+	CacheBucket string
+
+	// DestinationPath is desired path of the file after it's copied from CacheBucket.
+	DestinationPath string
 }
 
-// Upload uploads static assets to Cloud Storage and returns uploaded file URLs.
-func Upload(fs afero.Fs, source, destination string, opts *UploadOpts) ([]string, error) {
+// UploadToCache uploads the file(s) at source to u's CacheBucket.
+// Returns a list of cached assets successfully uploaded to the cache and an error, if any.
+func (u *Uploader) UploadToCache(source, dest string, opts *UploadOpts) ([]Cached, error) {
 	matcher := buildCompositeMatchers(buildReincludeMatchers(opts.Reincludes), buildExcludeMatchers(opts.Excludes))
-	var urls []string
-	info, err := fs.Stat(source)
-	if err != nil {
-		return nil, fmt.Errorf("get stat for file %q: %w", source, err)
-	}
-	paths := []string{source}
-	if info.IsDir() {
-		files, err := afero.ReadDir(fs, source)
-		if err != nil {
-			return nil, fmt.Errorf("read directory %q: %w", source, err)
-		}
-		paths = make([]string, len(files))
-		for i, f := range files {
-			paths[i] = filepath.Join(source, f.Name())
-		}
-	} else if destination == "" { // only applies to files, not folders
-		destination = source
+
+	var assets []Cached
+	if err := afero.Walk(u.FS, source, u.walkFn(source, dest, opts.Recursive, matcher, &assets)); err != nil {
+		return nil, fmt.Errorf("walk the file tree rooted at %q: %w", source, err)
 	}
-	for _, path := range paths {
-		if err := afero.Walk(fs, path, walkFn(source, destination, opts.Recursive, fs, opts.UploadFn, &urls, matcher)); err != nil {
-			return nil, fmt.Errorf("walk the file tree rooted at %q: %w", source, err)
-		}
+
+	if err := u.uploadAssets(assets); err != nil {
+		return nil, err
 	}
-	return urls, nil
+
+	return assets, nil
 }
 
-func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadFunc, urlsPtr *[]string, matcher filepathMatcher) filepath.WalkFunc {
+func (u *Uploader) walkFn(sourcePath, destPath string, recursive bool, matcher filepathMatcher, assets *[]Cached) filepath.WalkFunc {
 	return func(path string, info fs.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
 		if info.IsDir() {
-			if !recursive {
+			if !recursive && path != sourcePath {
```

Comment:
why do we make this change?

Reply:
since i removed the pre-processing of directories [here](https://github.com/aws/copilot-cli/pull/4722/files#diff-befc0e16cf8c6a7ad74caee13773fd4b095ba8bea5170dd1e39f6e3384e0667fL37-L44), we need to go through all paths in a directory if it is the directory provided.

example: for source `./files`
on first call to the func returned by `walkFn()`, path == sourcePath == "./files". we don't want to skip that directory, because it's the actual directory that was passed in.

Reply:
ah i see. ok then why do we remove the pre-processing of directories 🤔 i feel like it was easier to follow

Reply:
it simplifies the handling of figuring out what the destination location should be in `walkFn` - previously, files configured were passed in to `walkFn` the exact same was a directory configurations, so it was impossible (difficult?) to figure out if `sourcePath` of `./files/index.html` meant dest should be `index.html` in the case of the configuration being a file, or dest should be `/files/index.html`, in the case of the configuration being `./files`. Does that make any sense?